src/v/datalake/tests/catalog_schema_manager_test.cc

```diff
@@ -0,0 +1,181 @@
+      , schema_mgr(catalog) {
+        set_expectations_and_listen({});
+    }
+    cloud_io::remote& remote() { return sr->remote.local(); }
+
+    void reset_field_ids(struct_type& type) {
+        chunked_vector<nested_field*> to_visit;
+        for (auto& f : std::ranges::reverse_view(type.fields)) {
+            to_visit.emplace_back(f.get());
+        }
+        while (!to_visit.empty()) {
+            auto* f = to_visit.back();
+            f->id = nested_field::id_t{0};
+            to_visit.pop_back();
+            std::visit(reverse_field_collecting_visitor{to_visit}, f->type);
+        }
+    }
+
+    void create_test_table() {
+        schema s{
+          .schema_struct = std::get<struct_type>(test_nested_schema_type()),
+          .schema_id = schema::id_t{1},
+          .identifier_field_ids{},
+        };
+        auto create_res
+          = catalog.create_table(table_ident, s, partition_spec{}).get();
+        ASSERT_FALSE(create_res.has_error());
+    }
+
+    ss::future<std::optional<schema>>
+    load_table_schema(const table_identifier& table_ident) {
+        auto load_res = catalog.load_table(table_ident).get();
+        if (!load_res.has_value()) {
+            co_return std::nullopt;
+        }
+        auto& table = load_res.value();
+        EXPECT_NE(table.current_schema_id, schema::unassigned_id);
+        auto schema_it = std::ranges::find(
+          table.schemas, table.current_schema_id, &schema::schema_id);
+        if (schema_it == table.schemas.end()) {
+            co_return std::nullopt;
+        }
+        co_return std::move(*schema_it);
+    }
+
+    std::unique_ptr<cloud_io::scoped_remote> sr;
+    filesystem_catalog catalog;
+    catalog_schema_manager schema_mgr;
+};
+
+TEST_F(CatalogSchemaManagerTest, TestCreateTable) {
+    auto type = std::get<struct_type>(test_nested_schema_type());
+    reset_field_ids(type);
+
+    // Trying to fill the field IDs requires querying the catalog, which will
+    // result in the table being created.
+    auto res = schema_mgr.get_registered_ids(model::topic{"foo"}, type).get();
+    ASSERT_FALSE(res.has_error());
+
+    auto table_ident = table_identifier{.ns = {"redpanda"}, .table = "foo"};
+    auto schema = load_table_schema(table_ident).get();
+    ASSERT_TRUE(schema.has_value());
+    EXPECT_EQ(type, schema->schema_struct);
+}
+
+TEST_F(CatalogSchemaManagerTest, TestFillFromExistingTable) {
+    create_test_table();
+    auto schema = load_table_schema(table_ident).get();
+    ASSERT_TRUE(schema.has_value());
+
+    // Even if the table already exists, we should be able to fill fields IDs
+    // without trouble.
+    auto type = std::get<struct_type>(test_nested_schema_type());
+    reset_field_ids(type);
+    auto res = schema_mgr.get_registered_ids(model::topic{"foo"}, type).get();
+    ASSERT_FALSE(res.has_error());
+    EXPECT_EQ(type, schema.value().schema_struct);
+}
+
+TEST_F(CatalogSchemaManagerTest, TestFillSubset) {
```

Comment:
Can we test this case for a nested structure?

Reply:
DOne